marlowe-dashboard-client/src/Capability/PlutusApps/MarloweApp.purs

```diff
@@ -38,48 +47,134 @@ import Plutus.V1.Ledger.Value (TokenName) as Back
   redeem plutusAppId marloweContractId tokenName pubKeyHash = do
+    reqId <- liftEffect genUUID
     let
-      payload :: MarloweParams Json./\ Back.TokenName Json./\ Back.PubKeyHash
-      payload = marloweContractId Json./\ toBack tokenName Json./\ toBack pubKeyHash
-    mutex <- asks $ view (_marloweAppEndpointMutex <<< _redeem)
-    -- TODO: we could later add a forkAff with a timer that unlocks this timer if we
-    --       dont get a response
-    liftAff $ AVar.take mutex
-    Contract.invokeEndpoint plutusAppId "redeem" payload
+      payload :: UUID Json./\ MarloweParams Json./\ Back.TokenName Json./\ Back.PubKeyHash
+      payload = reqId Json./\ marloweContractId Json./\ toBack tokenName Json./\ toBack pubKeyHash
+    invokeMutexedEndpoint plutusAppId reqId "redeem" _redeem payload
 
 createEndpointMutex :: Effect EndpointMutex
 createEndpointMutex = do
   create <- EAVar.empty
   applyInputs <- EAVar.empty
   redeem <- EAVar.empty
-  pure { create, applyInputs, redeem }
+  requests <- EAVar.new mempty
+  pure { create, applyInputs, redeem, requests }
+
+invokeMutexedEndpoint ::
+  forall payload m env.
+  Encode payload =>
+  MonadAff m =>
+  ManageContract m =>
+  MonadAsk (MarloweAppEndpointMutexEnv env) m =>
+  PlutusAppId ->
+  UUID ->
+  String ->
+  Lens' EndpointMutex (AVar Unit) ->
+  payload ->
+  m (AjaxResponse Unit)
+invokeMutexedEndpoint plutusAppId reqId endpointName _endpointMutex payload = do
+  -- There are three mutex involved in this operation:
+  --   * The endpointMutex help us avoid making multiple requests to the same endpoint if its not
+  --     available.
+  --   * The global requestMutex help us manage the request queue concurrently.
+  --   * For each request in the queue, we store one mutex that represents the response.
+  endpointMutex <- asks $ view (_marloweAppEndpointMutex <<< _endpointMutex)
+  requestMutex <- asks $ view (_marloweAppEndpointMutex <<< _requests)
+  -- First we need to see if the endpoint is available to make a request
+  -- TODO: we could later add a forkAff with a timer that unlocks this timer if we
+  --       dont get a response
+  -- TODO: We could change this take for a read and listen for a 500 error with EndpointUnavailabe
+  --       to retry and take it (instead of preemptively taking it).
+  liftAff $ AVar.take endpointMutex
+  result <- Contract.invokeEndpoint plutusAppId endpointName payload
+  -- After making the request, we lock on the global request array so we can add
+  -- a new request id with its corresponding mutex.
+  -- TODO: It would be nice if the invokeEndpoint returns a requestId instead of having to generate
+  --       one from the FE.
+  liftAff do
+    requests <- AVar.take requestMutex
+    newRequestMutex <- AVar.empty
+    let
+      -- We add new requests to the begining of the array and remove them from the end.
+      newRequests = take 15 $ (reqId /\ newRequestMutex) : requests
```

Comment:
Where is the 15 coming from here? Could you give a name to this value so its purpose is more clear?

Reply:
sure, I'll add a name. The value is almost random, but I put the "meaning" in the reason of using an Array instead of a Map. I don't want to store ALL requests, just the last 10/15/20/50.